cexec/main.go

```diff
@@ -1,111 +1,202 @@
 		if err != nil {
-			log.Fatalf("Error changing root to [%s]", mountAction)
+			return fmt.Errorf("error changing root to [%s], error: %w", mountAction, err)
 		}
 	}
 
-	if defaultInterpreter != "" {
+	if s.defaultInterpreter != "" {
 		// Split the interpreter by space, in the event that the default intprepretter has flags.
-		di := strings.Split(defaultInterpreter, " ")
+		di := strings.Split(s.defaultInterpreter, " ")
 		if len(di) == 0 {
-			log.Fatalf("Error parsing [\"DEFAULT_INTERPETER\"] [%s]\n", defaultInterpreter)
+			return fmt.Errorf("error parsing [\"DEFAULT_INTERPRETER\"] [%s]", s.defaultInterpreter)
 		}
 		// Look for default shell intepreter
-		_, err = os.Stat(di[0])
-		if os.IsNotExist(err) {
-			log.Fatalf("Unable to find the [\"DEFAULT_INTERPETER\"] [%s], check chroot and interpreter path", defaultInterpreter)
+		if _, err := os.Stat(di[0]); os.IsNotExist(err) {
+			return fmt.Errorf("unable to find the [\"DEFAULT_INTERPRETER\"] [%s], check chroot and interpreter path", s.defaultInterpreter)
 		}
-		di = append(di, cmdLine)
+		di = append(di, s.cmdLine)
 		cmd := exec.Command(di[0], di[1:]...)
 		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
 		debugCMD := fmt.Sprintf("%s %v", di[0], di[1:])
-		err = cmd.Start()
-		if err != nil {
-			log.Fatalf("Error starting [%s] [%v]", debugCMD, err)
+		if err := cmd.Start(); err != nil {
+			return fmt.Errorf("error starting [%s] [%w]", debugCMD, err)
 		}
-		err = cmd.Wait()
-		if err != nil {
-			log.Fatalf("Error running [%s] [%v]", debugCMD, err)
+		if err := cmd.Wait(); err != nil {
+			return fmt.Errorf("error running [%s] [%w]", debugCMD, err)
 		}
 	} else {
 		// Format the cmdLine string into separate execution tasks
-		commandLines := strings.Split(cmdLine, ";")
+		commandLines := strings.Split(s.cmdLine, ";")
 		for x := range commandLines {
 			command := strings.Split(commandLines[x], " ")
 			cmd := exec.Command(command[0], command[1:]...)
 			cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
 			debugCMD := fmt.Sprintf("%s %v", command[0], command[1:])
-			err = cmd.Start()
-			if err != nil {
-				log.Fatalf("Error starting [%s] [%v]", debugCMD, err)
+			if err := cmd.Start(); err != nil {
+				return fmt.Errorf("error starting [%s] [%w]", debugCMD, err)
 			}
-			err = cmd.Wait()
-			if err != nil {
-				log.Fatalf("Error running [%s] [%v]", debugCMD, err)
+			if err := cmd.Wait(); err != nil {
+				return fmt.Errorf("error running [%s] [%w]", debugCMD, err)
 			}
 		}
 	}
 
-	if chroot != "" {
-		err = exitChroot()
-		if err != nil {
-			log.Errorf("Error exiting root from [%s], execution continuing", mountAction)
+	if s.chroot != "" {
+		if err := exitChroot(); err != nil {
+			log.Info("error exiting chroot, execution continuing", "chrootMountPoint", mountAction, "error", err)
+		}
+		if err := s.umountSpecialDirs(); err != nil {
+			log.Error("error unmounting special directories", "error", err)
 		}
 	}
```

Comment:
Given we're just logging these errors, would it make sense to defer this block and locate it next to the chroot previous `if chroot != ""`?

Reply:
sure thing.